src/components/Search/SearchSelectionListWithHeader.tsx

```diff
@@ -0,0 +1,124 @@
+import type {ForwardedRef} from 'react';
+import React, {forwardRef, useEffect, useState} from 'react';
+import SelectionList from '@components/SelectionList';
+import type {BaseSelectionListProps, ReportListItemType, SelectionListHandle, TransactionListItemType} from '@components/SelectionList/types';
+import * as SearchUtils from '@libs/SearchUtils';
+import CONST from '@src/CONST';
+import type {SearchDataTypes, SearchQuery, SelectedTransactions} from '@src/types/onyx/SearchResults';
+import SearchPageHeader from './SearchPageHeader';
+
+type SearchSelectionListWithHeaderProps = Omit<BaseSelectionListProps<ReportListItemType | TransactionListItemType>, 'onSelectAll' | 'onCheckboxPress' | 'sections'> & {
+    query: SearchQuery;
+    hash: number;
+    data: TransactionListItemType[] | ReportListItemType[];
+    searchType: SearchDataTypes;
+};
+
+function SearchSelectionListWithHeader({ListItem, onSelectRow, query, hash, data, searchType, ...props}: SearchSelectionListWithHeaderProps, ref: ForwardedRef<SelectionListHandle>) {
+    const [selectedItems, setSelectedItems] = useState<SelectedTransactions>({});
+
+    const clearSelectedItems = () => setSelectedItems({});
+
+    useEffect(() => {
+        clearSelectedItems();
+    }, [hash]);
+
+    const toggleTransaction = (item: TransactionListItemType | ReportListItemType) => {
+        if (SearchUtils.isTransactionListItemType(item)) {
+            if (!item.keyForList) {
+                return;
+            }
+
+            setSelectedItems((prev) => {
+                if (prev[item.keyForList]?.isSelected) {
+                    const {[item.keyForList]: omittedTransaction, ...transactions} = prev;
+                    return transactions;
+                }
+                return {...prev, [item.keyForList]: {isSelected: true, canDelete: item.canDelete, action: item.action}};
+            });
+
+            return;
+        }
+
+        if (item.transactions.every((transaction) => selectedItems[transaction.keyForList]?.isSelected)) {
+            const reducedSelectedItems: SelectedTransactions = {...selectedItems};
+
+            item.transactions.forEach((transaction) => {
+                delete reducedSelectedItems[transaction.keyForList];
+            });
+
+            setSelectedItems(reducedSelectedItems);
+            return;
+        }
+
+        setSelectedItems({
+            ...selectedItems,
+            ...Object.fromEntries(item.transactions.map((transaction) => [transaction.keyForList, {isSelected: true, canDelete: transaction.canDelete, action: transaction.action}])),
+        });
+    };
+
+    const toggleAllTransactions = () => {
+        const areReportItems = searchType === CONST.SEARCH.DATA_TYPES.REPORT;
```

Comment:
This name is unclear. What does it mean? "are there any report items"? "are all items report items"?
Please rename to something clearer